File: shipvision/mtmc/clustering/agglomerative.py
"""Average-linkage agglomerative clustering with a distance cut.

Four words carry the whole algorithm, and each of them is a decision:

**Agglomerative**, because the number of identities in front of a camera group at one instant
is exactly what is being asked and cannot be supplied.

**Average linkage**, because single linkage chains — A near B, B near C, so A, B and C are one
person even though A and C are strangers — and complete linkage refuses to admit a third view
of an identity whose worst pairing is mediocre, which is what a partially-occluded crop always
is. Average is also what makes the ``NEVER_MERGE`` sentinel work: one forbidden pair drags a
candidate merge's mean distance to ~50 000, so the group is never formed.

**A distance cut**, not a cluster count: ``fcluster(..., criterion="distance")``.

**On a precomputed matrix**, because the evidence combined in that matrix — cosine appearance,
a ground-plane veto, a same-camera exclusion — is not a metric embedding of anything and there
are no coordinates to hand a clusterer instead.

This is a ~15-line call into ``scipy``. The reference vendors 2 400 lines of hand-written C++
reciprocal-agglomerative clustering to do it, which is the exact shape of code the ponytail
principle exists to prevent: a well-optimised library primitive, reimplemented, now needing
its own tests and its own maintainer.
"""

from __future__ import annotations

from typing import Any

import numpy as np

from shipvision.errors import BackendUnavailableError, ConfigurationError
from shipvision.mtmc.clustering.base import CLUSTERERS, BaseClusterer
from shipvision.registry import PYTHON

__all__ = ["AgglomerativeClusterer"]


def _load_scipy() -> tuple[Any, Any, Any]:
    try:
        from scipy.cluster.hierarchy import fcluster, linkage
        from scipy.spatial.distance import squareform
    except ImportError as exc:  # pragma: no cover - depends on the host
        raise BackendUnavailableError(
            "agglomerative clustering needs scipy; install the 'solvers' extra "
            "(pip install 'shipvision[solvers]')"
        ) from exc
    return linkage, fcluster, squareform


@CLUSTERERS.register("agglomerative", backend=PYTHON, aliases=("average_linkage", "aic"))
class AgglomerativeClusterer(BaseClusterer):
    """Hierarchical clustering cut at a distance threshold."""

    def __init__(self, *, distance_threshold: float = 0.14) -> None:
        """
        Args:
            distance_threshold: the cut. In the same units as the matrix, so with the
                appearance builder it is ``1 - cosine_similarity`` — 0.14 meaning "group
                things that are at least 0.86 alike", which is the reference's production
                value and deliberately the same number as its appearance threshold. The two
                doing the same work is not redundant: the appearance threshold removes weak
                *pairs* before linkage, and this one bounds the *average* over a group, so
                without the first a group of mediocre pairs still averages under the cut.
        """
        if distance_threshold <= 0.0:
            raise ConfigurationError(
                f"distance_threshold must be positive, got {distance_threshold}"
            )
        self.distance_threshold = float(distance_threshold)

    def fit_predict(self, distances: np.ndarray) -> np.ndarray:
        matrix = self.check_matrix(distances)
        count = matrix.shape[0]
        if count == 0:
            return np.zeros(0, dtype=np.int32)
        if count == 1:
            return np.zeros(1, dtype=np.int32)

        linkage, fcluster, squareform = _load_scipy()
        # Symmetrise and zero the diagonal before condensing. squareform's own checks are
        # exact — a 1e-16 asymmetry out of BLAS fails them — and with checks off it silently
        # keeps the upper triangle, so doing it explicitly is what makes the result the
        # matrix the builder meant rather than half of it.
        symmetric = 0.5 * (matrix + matrix.T)
        np.fill_diagonal(symmetric, 0.0)
        condensed = squareform(symmetric, checks=False)
        tree = linkage(condensed, method="average")
        labels = fcluster(tree, t=self.distance_threshold, criterion="distance")
        # scipy labels from 1; zero-based keeps "label 0" from meaning two things depending
        # on which library produced it.
        return (np.asarray(labels, dtype=np.int32) - 1).astype(np.int32)

    def __repr__(self) -> str:
        return (
            f"<AgglomerativeClusterer distance_threshold={self.distance_threshold} "
            f"backend={self.backend}>"
        )

File: csrc/shipvision/mtmc/clustering/agglomerative.cpp
#include "shipvision/mtmc/clustering/agglomerative.h"

#include <cmath>
#include <limits>
#include <stdexcept>
#include <string>

namespace shipvision::mtmc {

    AgglomerativeClusterer::AgglomerativeClusterer(double distance_threshold)
        : distance_threshold_(distance_threshold) {
        if (!(distance_threshold > 0.0)) {
            throw std::invalid_argument("distance_threshold must be positive, got " +
                                        std::to_string(distance_threshold));
        }
    }

    std::vector<int> AgglomerativeClusterer::fit_predict(const double* distances, int n) const {
        if (n <= 0)
            return {};
        if (n == 1)
            return {0};

        const size_t side = static_cast<size_t>(n);
        std::vector<double> working(side * side);
        for (size_t i = 0; i < side; ++i) {
            for (size_t j = 0; j < side; ++j) {
                const double value = distances[i * side + j];
                if (!std::isfinite(value)) {
                    throw std::invalid_argument(
                        "the distance matrix contains inf or NaN. Use the finite NEVER_MERGE "
                        "sentinel for pairs that must not be grouped: hierarchical clustering "
                        "cannot consume non-finite distances, and average linkage turns them "
                        "into NaN rather than into a refusal");
                }
                working[i * side + j] = value;
            }
        }
        // Symmetrise and zero the diagonal before anything reads a pair. Both are what the
        // Python side does before handing scipy a condensed matrix, and for the same reason:
        // `squareform`'s tolerance for asymmetry is zero, and with its checks off it silently
        // keeps the upper triangle — so half a matrix gets clustered and the result is plausible.
        for (size_t i = 0; i < side; ++i) {
            for (size_t j = i + 1; j < side; ++j) {
                const double mean = 0.5 * (working[i * side + j] + working[j * side + i]);
                working[i * side + j] = mean;
                working[j * side + i] = mean;
            }
            working[i * side + i] = 0.0;
        }

        std::vector<int> parent(side);
        std::vector<double> weight(side, 1.0);
        std::vector<unsigned char> active(side, 1);
        for (size_t index = 0; index < side; ++index)
            parent[index] = static_cast<int>(index);

        for (int merges = 0; merges + 1 < n; ++merges) {
            double best = std::numeric_limits<double>::infinity();
            int left = -1;
            int right = -1;
            for (int i = 0; i < n; ++i) {
                if (!active[static_cast<size_t>(i)])
                    continue;
                for (int j = i + 1; j < n; ++j) {
                    if (!active[static_cast<size_t>(j)])
                        continue;
                    const double value = working[static_cast<size_t>(i) * side + j];
                    if (value < best) {
                        best = value;
                        left = i;
                        right = j;
                    }
                }
            }
            // `<=`, not `<`. scipy's `fcluster(criterion="distance")` keeps every merge whose
            // cophenetic distance is at most t, and average linkage is monotone, so stopping the
            // greedy sequence at the first merge above t reproduces exactly that cut. A strict
            // comparison here would split a group whose members sit exactly on the threshold —
            // which is what a threshold copied from a tuning run is made of.
            if (left < 0 || !(best <= distance_threshold_))
                break;

            // Lance-Williams for average linkage (UPGMA): the merged cluster's distance to a
            // third is the size-weighted mean of its parts'. Weighted by cluster SIZE, not by a
            // plain average of the two — the unweighted variant (WPGMA) is a different algorithm
            // that lets a singleton outvote a group of five, and both are called "average".
            const double left_weight = weight[static_cast<size_t>(left)];
            const double right_weight = weight[static_cast<size_t>(right)];
            const double total = left_weight + right_weight;
            for (int k = 0; k < n; ++k) {
                if (k == left || k == right || !active[static_cast<size_t>(k)])
                    continue;
                const double merged =
                    (left_weight * working[static_cast<size_t>(left) * side + k] +
                     right_weight * working[static_cast<size_t>(right) * side + k]) /
                    total;
                working[static_cast<size_t>(left) * side + k] = merged;
                working[static_cast<size_t>(k) * side + left] = merged;
            }
            weight[static_cast<size_t>(left)] = total;
            active[static_cast<size_t>(right)] = 0;
            parent[static_cast<size_t>(right)] = left;
        }

        std::vector<int> labels(side, -1);
        int next_label = 0;
        for (int index = 0; index < n; ++index) {
            int root = index;
            while (parent[static_cast<size_t>(root)] != root)
                root = parent[static_cast<size_t>(root)];
            // Numbered by first appearance, so that "these two implementations agree" is one
            // array comparison rather than a partition isomorphism.
            if (labels[static_cast<size_t>(root)] < 0)
                labels[static_cast<size_t>(root)] = next_label++;
            labels[static_cast<size_t>(index)] = labels[static_cast<size_t>(root)];
        }
        return labels;
    }

}  // namespace shipvision::mtmc
